File: sdk/go/provider/getnada.go
package provider

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"math/rand"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const getnadaBase = "https://getnada.net/api"

type getnadaDomainsResponse struct {
	Domains []string `json:"domains"`
}

type getnadaOpenResponse struct {
	InboxID     string `json:"inboxId"`
	Token       string `json:"token"`
	ActiveUntil string `json:"activeUntil"`
	Recipient   string `json:"recipient"`
}

type getnadaMessagesResponse struct {
	Messages []map[string]any `json:"messages"`
}

type getnadaDetailResponse struct {
	Message map[string]any `json:"message"`
}

func getnadaRandomLocal() string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	var b strings.Builder
	b.WriteString("sdk")
	for i := 0; i < 16; i++ {
		b.WriteByte(chars[rand.Intn(len(chars))])
	}
	return b.String()
}

func getnadaJSON(method, u string, body []byte, out any) error {
	req, err := http.NewRequest(method, u, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/json")
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	resp, err := HTTPClient().Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("getnada: http %d", resp.StatusCode)
	}
	if out == nil {
		return nil
	}
	return json.Unmarshal(raw, out)
}

func getnadaPickDomain() (string, error) {
	var data getnadaDomainsResponse
	if err := getnadaJSON(http.MethodGet, getnadaBase+"/public/domains", nil, &data); err != nil {
		return "", err
	}
	for _, domain := range data.Domains {
		if strings.TrimSpace(domain) == "getnada.net" {
			return "getnada.net", nil
		}
	}
	for _, domain := range data.Domains {
		if d := strings.TrimSpace(domain); d != "" {
			return d, nil
		}
	}
	return "", fmt.Errorf("getnada: no domain available")
}

func getnadaString(v any) string {
	switch x := v.(type) {
	case string:
		return x
	case nil:
		return ""
	default:
		return fmt.Sprintf("%v", x)
	}
}

func getnadaFlattenMessage(raw map[string]any, recipient string) map[string]any {
	out := make(map[string]any, len(raw)+6)
	for k, v := range raw {
		out[k] = v
	}
	out["id"] = raw["id"]
	if v, ok := raw["from_addr"]; ok {
		out["from"] = v
	} else if v, ok := raw["from"]; ok {
		out["from"] = v
	}
	if v, ok := raw["to"]; ok && getnadaString(v) != "" {
		out["to"] = v
	} else {
		out["to"] = recipient
	}
	if v, ok := raw["text_plain"]; ok {
		out["text"] = v
	} else if v, ok := raw["text"]; ok {
		out["text"] = v
	}
	if v, ok := raw["html_sanitized"]; ok {
		out["html"] = v
	} else if v, ok := raw["html"]; ok {
		out["html"] = v
	}
	if v, ok := raw["read_at"]; ok {
		out["read"] = getnadaString(v) != ""
	}
	return out
}

func GetnadaGenerate() (*CreatedMailbox, error) {
	domain, err := getnadaPickDomain()
	if err != nil {
		return nil, err
	}
	requested := getnadaRandomLocal() + "@" + domain
	body, err := json.Marshal(map[string]string{"email": requested})
	if err != nil {
		return nil, err
	}
	var data getnadaOpenResponse
	if err := getnadaJSON(http.MethodPost, getnadaBase+"/inbox/open", body, &data); err != nil {
		return nil, err
	}
	token := strings.TrimSpace(data.Token)
	email := strings.TrimSpace(data.Recipient)
	if email == "" {
		email = requested
	}
	if token == "" || email == "" || !strings.Contains(email, "@") {
		return nil, fmt.Errorf("getnada: invalid open response")
	}
	return &CreatedMailbox{
		Channel:   "getnada",
		Email:     email,
		Token:     token,
		ExpiresAt: data.ActiveUntil,
	}, nil
}

func getnadaFetchDetail(token, messageID string) (map[string]any, error) {
	q := url.Values{}
	q.Set("id", messageID)
	q.Set("token", token)
	var data getnadaDetailResponse
	if err := getnadaJSON(http.MethodGet, getnadaBase+"/inbox/message?"+q.Encode(), nil, &data); err != nil {
		return nil, err
	}
	if data.Message == nil {
		return nil, fmt.Errorf("getnada: empty message detail")
	}
	return data.Message, nil
}

func GetnadaGetEmails(token, email string) ([]NormEmail, error) {
	auth := strings.TrimSpace(token)
	address := strings.TrimSpace(email)
	if auth == "" {
		return nil, fmt.Errorf("getnada: empty token")
	}
	if address == "" {
		return nil, fmt.Errorf("getnada: empty email")
	}

	q := url.Values{}
	q.Set("token", auth)
	var data getnadaMessagesResponse
	if err := getnadaJSON(http.MethodGet, getnadaBase+"/inbox/messages?"+q.Encode(), nil, &data); err != nil {
		return nil, err
	}
	out := make([]NormEmail, 0, len(data.Messages))
	for _, row := range data.Messages {
		raw := row
		if id := strings.TrimSpace(getnadaString(row["id"])); id != "" {
			if detail, err := getnadaFetchDetail(auth, id); err == nil {
				raw = detail
			}
		}
		out = append(out, NormalizeMap(getnadaFlattenMessage(raw, address), address))
	}
	return out, nil
}

File: sdk/npm/src/providers/getnada.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'getnada';
const API_BASE = 'https://getnada.net/api';
const DOMAIN_LABEL_RE = /^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$/i;

interface DomainsResponse {
  domains?: string[];
}

interface OpenResponse {
  inboxId?: string;
  token?: string;
  activeUntil?: string;
  recipient?: string;
}

interface MessagesResponse {
  messages?: Array<Record<string, unknown>>;
}

interface MessageDetailResponse {
  message?: Record<string, unknown>;
}

function randomLocal(): string {
  const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
  let out = 'sdk';
  for (let i = 0; i < 16; i++) out += chars[Math.floor(Math.random() * chars.length)];
  return out;
}

async function fetchJSON<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetchWithTimeout(url, {
    ...init,
    headers: {
      Accept: 'application/json',
      ...(init?.headers || {}),
    },
  });
  if (!response.ok) {
    throw new Error(`getnada http ${response.status}`);
  }
  return response.json() as Promise<T>;
}

function flattenMessage(raw: Record<string, unknown>, recipient: string): Record<string, unknown> {
  return {
    ...raw,
    id: raw.id,
    from: raw.from_addr || raw.from,
    to: raw.to || recipient,
    text: raw.text_plain || raw.text,
    html: raw.html_sanitized || raw.html,
    read: raw.read_at ? true : raw.read,
  };
}

async function pickDomain(preferred?: string | null): Promise<string> {
  const data = await fetchJSON<DomainsResponse>(`${API_BASE}/public/domains`);
  const domains = (Array.isArray(data.domains) ? data.domains : [])
    .map((domain) => String(domain || '').trim().toLowerCase())
    .filter(isMailDomain);
  const wanted = String(preferred || '').trim().replace(/^@/, '').toLowerCase();
  if (wanted) {
    const exact = domains.find((domain) => domain === wanted);
    if (!exact) throw new Error(`getnada: domain not available: ${wanted}`);
    return exact;
  }
  return domains.find((domain) => domain === 'getnada.net') ?? domains[0] ?? '';
}

function isMailDomain(domain: string): boolean {
  if (!domain || domain.length > 253 || domain.includes('..')) return false;
  const labels = domain.split('.');
  return labels.length >= 2 && labels.every((label) => DOMAIN_LABEL_RE.test(label));
}

export async function generateEmail(domain?: string | null, channel: Channel = CHANNEL): Promise<InternalEmailInfo> {
  const selectedDomain = await pickDomain(domain);
  if (!selectedDomain) throw new Error('getnada: no domain available');
  const requested = `${randomLocal()}@${selectedDomain}`;
  const data = await fetchJSON<OpenResponse>(`${API_BASE}/inbox/open`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ email: requested }),
  });

  const token = String(data.token || '').trim();
  const email = String(data.recipient || requested).trim();
  if (!token || !email || !email.includes('@')) {
    throw new Error('getnada: invalid open response');
  }
  return {
    channel,
    email,
    token,
    expiresAt: data.activeUntil,
    createdAt: undefined,
  };
}

async function fetchDetail(token: string, messageId: string): Promise<Record<string, unknown> | null> {
  const url = `${API_BASE}/inbox/message?id=${encodeURIComponent(messageId)}&token=${encodeURIComponent(token)}`;
  try {
    const data = await fetchJSON<MessageDetailResponse>(url);
    return data.message && typeof data.message === 'object' ? data.message : null;
  } catch {
    return null;
  }
}

export async function getEmails(token: string, email: string): Promise<Email[]> {
  const auth = String(token || '').trim();
  if (!auth) throw new Error('getnada: empty token');
  const address = String(email || '').trim();
  if (!address) throw new Error('getnada: empty email');

  const data = await fetchJSON<MessagesResponse>(`${API_BASE}/inbox/messages?token=${encodeURIComponent(auth)}`);
  const rows = Array.isArray(data.messages) ? data.messages : [];

  const emails: Email[] = [];
  for (const row of rows) {
    const id = row.id == null ? '' : String(row.id);
    const detail = id ? await fetchDetail(auth, id) : null;
    emails.push(normalizeEmail(flattenMessage(detail || row, address), address));
  }
  return emails;
}
